packages/dd-trace/src/appsec/iast/analyzers/nosql-injection-mongodb-analyzer.js

```diff
@@ -0,0 +1,167 @@
+      const store = storage.getStore()
+      if (store && !store.nosqlAnalyzed && filters?.length) {
+        filters.forEach(filter => {
+          this.analyze({ filter }, store)
+        })
+      }
+    })
+
+    this.addSub('datadog:mongoose:model:filter:start', ({ filters }) => {
+      const store = storage.getStore()
+      if (!store) return
+
+      if (filters?.length) {
+        filters.forEach(filter => {
+          this.analyze({ filter }, store)
+        })
+      }
+
+      storage.enterWith({ ...store, nosqlAnalyzed: true, mongooseParentStore: store })
+    })
+
+    this.addSub('datadog:mongoose:model:filter:finish', () => {
+      const store = storage.getStore()
+      if (store?.mongooseParentStore) {
+        storage.enterWith(store.mongooseParentStore)
+      }
+    })
+  }
+
+  configureSanitizers () {
+    this.addNotSinkSub('datadog:express-mongo-sanitize:filter:finish', ({ sanitizedProperties, req }) => {
+      const store = storage.getStore()
+      const iastContext = getIastContext(store)
+
+      if (iastContext) { // do nothing if we are not in an iast request
+        sanitizedProperties.forEach(key => {
+          iterateObjectStrings(req[key], function (value, levelKeys) {
+            if (typeof value === 'string') {
+              let parentObj = req[key]
+              const levelsLength = levelKeys.length
+
+              for (let i = 0; i < levelsLength; i++) {
+                const currentLevelKey = levelKeys[i]
+
+                if (i === levelsLength - 1) {
+                  parentObj[currentLevelKey] = addSecureMark(iastContext, value, MONGODB_NOSQL_SECURE_MARK)
+                } else {
+                  parentObj = parentObj[currentLevelKey]
+                }
+              }
+            }
+          })
+        })
+      }
+    })
+
+    this.addNotSinkSub('datadog:express-mongo-sanitize:sanitize:finish', ({ sanitizedObject }) => {
+      const store = storage.getStore()
+      const iastContext = getIastContext(store)
+
+      if (iastContext) { // do nothing if we are not in an iast request
+        iterateObjectStrings(sanitizedObject, function (value, levelKeys, parent, lastKey) {
+          try {
+            parent[lastKey] = addSecureMark(iastContext, value, MONGODB_NOSQL_SECURE_MARK)
+          } catch {
+            // if it is a readonly property, do nothing
+          }
+        })
+      }
+    })
+
+    this.addNotSinkSub('datadog:mongoose:sanitize-filter:finish', ({ sanitizedObject }) => {
+      this.sanitizedObjects.add(sanitizedObject)
+    })
+  }
+
+  _isVulnerable (value, iastContext) {
+    if (value?.filter && iastContext) {
+      let isVulnerable = false
+
+      if (this.sanitizedObjects.has(value.filter)) {
+        return false
+      }
+
+      const rangesByKey = {}
+      const allRanges = []
+
+      iterateObjectStrings(value.filter, function (val, nextLevelKeys) {
+        const ranges = getRanges(iastContext, val)
```

Comment:
There is a false positive related to how `getRanges` is working here.

Assume the following application code:

```
const name = req.body.username;
const pass = req.body.password;
const query = {
    username: '',
    password: ''
}
User.find(query, function (err, user) { ..
```
If any of the tainted values is an empty string, this will be detected as a NoSQLi even though it is completely safe (no values coming from the user input are being used in the query).

The reason why is getRanges returns ranges that matches the empty string from query properties.

Reply:
Good catch Carles,
We've talked about it, and it looks like general issue that we will fix in a different PR.